### ISSUE TITLE
Once conch.ssh depend on PyCA cryptography look at replacing gmpy

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8079]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#8079|
|Type|enhancement|
|Created|2015-11-02 07:43:51Z|
|Branch|https://github.com/twisted/twisted/tree/8079-no-gmpy|

Once cryptography is merged I will try to look at gmpy... and see if it can be replaced with python cryptography.

Check that we still get a good performance... or maybe still use gmpy if it is available but also try python cryptography, if it has a better performance than the build it version.

Attachments:

* [timing.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/87e/87efc933f93946ad0d85188bdc2eea5511dba634/12f53210e0ad0acdaa0a27f481f1d8bc1bbcb7db.py) (1576 bytes) - added by markrwilliams on 2016-09-18 07:57:21Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8079 8079
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__None None
branch__8079_no_gmpy 8079-no-gmpy
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__None None
time__1446450231635011 1446450231635011
changetime__1477011447504199 1477011447504199
version__None None
owner__hawkowl hawkowl
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="daira's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|daira commented|
|-|-|

A reason to avoid gmpy is that it's unmaintained (the replacement is gmpy2) and does not have binary wheels on PyPI -- see https://tahoe-lafs.org/trac/tahoe-lafs/ticket/2094#comment:3

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

It's not installed on any of our builders, and if you want speed, use PyPy -- the Cryptography port means that conch now works fine on it.

We should remove it.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

PR: https://github.com/twisted/twisted/pull/540

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_7"></a>|@markrwilliams commented|
|-|-|

I'm worried that using `pow` for modular exponentiation during Diffie Hellman key exchange opens conch up for timing attacks.  I'm going off of [Kocher's 1996 "Timing Attacks" paper](http://courses.csail.mit.edu/6.857/2006/handouts/TimingAttacks.pdf).  While I'm sure it no longer represents the state of the art, it does say that reduction in modular multiplication results in exploitable timing variations.  I'm pretty sure that `pow` uses modular reduction for bignums (FIVEARY_CUTOFF is 8):

https://github.com/python/cpython/blob/2.7/Objects/longobject.c#L3502-L3560
https://github.com/python/cpython/blob/3.5/Objects/longobject.c#L3915-L3973

Obviously I'm not great at math, so I wrote a program that measures the time taken to do DH exponentiation via `cryptography`'s bindings to OpenSSL's bignum code or Python's `pow`.  I've attached it as `timing.py`.

Here are the results (`python` means `pow` and `openssl` means OpenSSL's bignums):

```
$ ministat -n python openssl
x python
+ openssl
    N           Min           Max        Median           Avg        Stddev
x 10000  0.0058999062   0.018929005  0.0061118603  0.0063149349 0.00057601654
+ 10000 0.00078105927  0.0015981197 0.00080513954 0.00082731056 8.3011552e-05
```

OpenSSL's bignums are both faster (lower median) and more consistent (smaller standard deviation.)  I think the former is nice and the latter is pretty important.

The downside is that cryptography doesn't expose easy-to-use bindings for OpenSSL's bignum code.  If what I'm saying sounds reasonable, I can look into those bindings.

Thoughts?

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban commented|
|-|-|

gmpy, at least on Ubuntu works... and is the fastest option available

I can do `pip install gmpy` and I get a big performance improvement.

I am using OpeSSH client as `ssh -oKexAlgorithms=diffie-hellman-group-exchange-sha256 -p 5022 user@localhost` and run a simple test using the simple ssh server example (https://gist.github.com/adiroiban/c0e309b6d58494b8ebea).

Some raw measurements on my laptop i7-4500U CPU @ 1.80GHz with Ubuntu 16.04:

* py27 without gmpy 3.5s
* py27 with gmpy 0.6s
* pypy 5.4.1 1.8s


My intention with this ticket was to check if we already have the equivalent methods in cryptography and try to use them instead of python code .

Since this is a soft dependency I think that we should remove it only if the openssl version is faster.

Installing gmpy on the Travis-ci should not be a complicated task.

Since pypy is still slower than py2.7 + gmpy, I am -1 on completely removing the gmpy support, as of now.

Leaving this in review to get feedback from more devs

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_9"></a>|@rodrigc set owner to @hawkowl|
|-|-|

https://github.com/twisted/twisted/pull/540#pullrequestreview-479322

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl commented|
|-|-|

So, markrwilliams: that sounds good, although I don't think it should block this ticket.

Adi: there is a bit more perf using gmpy for the initial connection; yes, but Twisted also shouldn't monkeypatch globals. People that want to accept the risk of gmpy possibly crashing and causing a DoS (which it's been known to do if you pass it the wrong things), in potentially random parts of their python, can copy out the old code and monkeypatch the globals themselves. Randomly changing behaviour in a potentially insecure and surprising way, to code that *happens* to import something from conch -- not even using it! -- I don't think is acceptable. 

If we still want gmpy in Conch, I think we should do it by integrating gmpy2, which publishes binary wheels, and implement it in a way that doesn't involve monkeypatching globals, instead explicitly passing in the ones they should use, or at least monkeypatching the module like the rest of the old code did.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_11"></a>|@markrwilliams commented|
|-|-|

Replying to [hawkowl](#note_10):
> So, markrwilliams: that sounds good, although I don't think it should block this ticket.
> 

Fair enough.  `gmpy` has enough problems that it should go.

> 
> If we still want gmpy in Conch, I think we should do it by integrating gmpy2

I don't think gmpy2 is worth the effort, because [OpenSSL actually has Diffie-Hellman calculation primitives](https://www.openssl.org/docs/manmaster/crypto/DH_compute_key.html).  (I was totally wrong about using BN_mod_exp before -- crypto is hard.)

I expect `DH_compute_key` to be comparable in speed and security to OpenSSH because OpenSSH itself uses it.  Unfortunately, `cryptography` doesn't expose a public API for this or `DH_generate_key` yet.  Once it does, we can revisit this in [#8831](https://github.com/twisted/twisted/issues/8831).

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_12"></a>|@rodrigc set status to `closed`|
|-|-|

In changeset 01611cba886c3f9946bafacc9c028ff1e2592f53
```
#!CommitTicketReference repository="" revision="01611cba886c3f9946bafacc9c028ff1e2592f53"
Merge pull request #540 from twisted/8079-no-gmpy

Author: hawkowl
Reviewer: rodrigc
Fixes: #8079 

Remove gmpy support
```
